### PR TITLE
Drop support for old-style buildpacks and plugins that do not retrieve config using config plugin

### LINKED ISSUE #2177: /app/.env not updated on dokku release / restart of apps but only during deploy/rebuild.
## Description of problem:

/app/.env that is populated during deploy and sourced by the app during each release is not updated during dokku release, causing the app to use the config var that was set prior to the app being deployed.

**Scenario:** Imagine a S3 bucket name being set using `dokku config:set $app BUCKET_NAME` prior to the app being deployed, and is supposed to be removed later, even though `dokku config:unset $app BUCKET_NAME` is issued the older bucket name still persists.

## Actual Results:

`/app/.env` is only updated during adding of BUILD ENV and is never updated during dokku release.

## Expected Results:

`/app/.env` is updated during [dokku release](https://github.com/dokku/dokku/blob/master/plugins/00_dokku-standard/subcommands/release#L21) similar to how it is updated while the [Build Env is added to the build environment](https://github.com/dokku/dokku/blob/master/plugins/build-env/pre-build-buildpack#L35) Herokuish [sources the file](https://github.com/gliderlabs/herokuish/blob/master/include/buildpack.bash#L93) causing the config vars to be loaded even though they are not set or are removed had they been set prior to the app being deployed.

## Debug Output:

- `uname -a`: Linux 3.13.0-48-generic #80-Ubuntu SMP Thu Mar 12 11:16:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
- `free -m`:   
```
total       used       free     shared    buffers     cached
Mem:          7479       7297        182         10       1355       3106
-/+ buffers/cache:       2835       4644
Swap:            0          0          0
```
- `docker version`:
```
Client:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.2
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   c3959b1
 Built:        Mon Feb 22 21:37:01 2016
 OS/Arch:      linux/amd64
```
- `docker -D info`:
```
Containers: XXX
 Running: XXX
 Paused: 0
 Stopped: XXX
Images: XXX
Server Version: 1.10.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 235
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 3.13.0-48-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.305 GiB
Name: XXXXXX.XXXXXX.XXXXXX
ID: XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX:XXXX
WARNING: No swap limit support
```
- `docker run -ti gliderlabs/herokuish:latest herokuish version`:
```
herokuish: 0.3.8
buildpacks:
  heroku-buildpack-clojure   v73
  heroku-buildpack-go        v22
  heroku-buildpack-gradle    v17
  heroku-buildpack-grails    v19
  heroku-buildpack-java      v42
  heroku-buildpack-multi     v1.0.0
  heroku-buildpack-nodejs    v87
  heroku-buildpack-php       v90
  heroku-buildpack-play      v24
  heroku-buildpack-python    v74
  heroku-buildpack-ruby      v141
  heroku-buildpack-scala     v64
  buildpack-nginx            v5
```
- `dokku version`:
`v0.4.13`
- `dokku plugin`:
```
  00_dokku-standard    0.4.0 enabled    dokku core standard plugin
  20_events            0.4.0 enabled    dokku core events logging plugin
  apps                 0.4.0 enabled    dokku core apps plugin
  apt                  0.4.0 enabled    Inject deb packages into dokku based on files in project
  backup               0.4.0 enabled    dokku core backup plugin
  build-env            0.4.0 enabled    dokku core build-env plugin
  certs                0.4.0 enabled    dokku core certificate management plugin
  checks               0.4.0 enabled    dokku core checks plugin
  common               0.4.0 enabled    dokku core common plugin
  config               0.4.0 enabled    dokku core config plugin
  docker-options       0.4.0 enabled    dokku core docker-options plugin
  domains              0.4.0 enabled    dokku core domains plugin
  elasticsearch        1.0.0 enabled    dokku elasticsearch service plugin
  enter                0.4.0 enabled    dokku core enter plugin
  git                  0.4.0 enabled    dokku core git plugin
  logs                 0.4.0 enabled    dokku core logs plugin
  mongo                1.0.0 enabled    dokku mongo service plugin
  mysql                1.0.0 enabled    dokku mysql service plugin
  named-containers     0.4.0.1 enabled    dokku core named containers plugin
  nginx-vhosts         0.4.0 enabled    dokku core nginx-vhosts plugin
  plugin               0.4.0 enabled    dokku core plugin plugin
  ps                   0.4.0 enabled    dokku core ps plugin
  rabbitmq             1.0.0 enabled    dokku rabbitmq service plugin
  redis                1.0.0 enabled    dokku redis service plugin
  secure-apps          1.0.0 enabled    secures an application using basic authentication
  shell                0.4.0 enabled    dokku core shell plugin
  tags                 0.4.0 enabled    dokku core tags plugin
  tar                  0.4.0 enabled    dokku core tar plugin
  volume               1.0.0 enabled    persistent storage plugin for dokku
```
How was dokku installed?: `From source`

## Steps to Reproduce:

1. Create an app.
    ```
    $ dokku apps:create configtest
    Creating configtest... done
    ```

2. Set config vars in app using `dokku config:set`

    ```
    $ dokku config:set configtest MYBUILDPACK='http://example.com/buildpack'
    -----> Setting config vars
           MYBUILDPACK: http://example.com/buildpack
    -----> Restarting app configtest
    App configtest has not been deployed
    ```

3. Deploy the app

    ```
    $git push dokku master
    Counting objects: 21, done.
    Delta compression using up to 8 threads.
    Compressing objects: 100% (14/14), done.
    Writing objects: 100% (21/21), 1.89 KiB | 0 bytes/s, done.
    Total 21 (delta 5), reused 13 (delta 2)
    -----> Cleaning up...
    -----> Building configtest from herokuish...
    -----> Injecting apt repositories and packages ...
    -----> Adding BUILD_ENV to build environment...
    -----> Python app detected
    -----> Installing runtime (python-2.7.11)
    -----> Installing dependencies with pip
           Collecting Flask==0.9 (from -r requirements.txt (line 1))
           Downloading Flask-0.9.tar.gz (481kB)
           Collecting Jinja2==2.6 (from -r requirements.txt (line 2))
           Downloading Jinja2-2.6.tar.gz (389kB)
           Collecting Werkzeug==0.8.3 (from -r requirements.txt (line 3))
           Downloading Werkzeug-0.8.3.tar.gz (1.1MB)
           Installing collected packages: Werkzeug, Jinja2, Flask
           Running setup.py install for Werkzeug
           Running setup.py install for Jinja2
           Running setup.py install for Flask
           Successfully installed Flask-0.9 Jinja2-2.6 Werkzeug-0.8.3
    remote: You are using pip version 7.1.2, however version 8.1.2 is available.
    remote: You should consider upgrading via the 'pip install --upgrade pip' command.
           
    -----> Discovering process types
           Procfile declares types -> web
    -----> Releasing configtest (dokku/configtest:latest)...
    -----> Deploying configtest (dokku/configtest:latest)...
    -----> DOKKU_SCALE file not found in app image. Generating one based on Procfile...
    -----> New DOKKU_SCALE file generated
    =====> web=1
    -----> Running pre-flight checks
           For more efficient zero downtime deployments, create a file CHECKS.
           See http://dokku.viewdocs.io/dokku/checks-examples.md for examples
           CHECKS file not found in container: Running simple container check...
    -----> Waiting for 5 seconds ...
    -----> Default container check successful!
    =====> configtest container output:
            * Running on http://0.0.0.0:5000/
    =====> end configtest container output
    -----> Running post-deploy
    =====> renaming container (b19c70a55154) trusting_booth to configtest.web.1
    -----> Creating new /home/dokku/configtest/VHOST...
    -----> Setting config vars
           DOKKU_NGINX_PORT: 80
    -----> No matching configured domains for configtest found in SSL certificate. Your app will show as insecure in a browser if accessed via SSL
    -----> Please add appropriate domains via the dokku domains command
    -----> Configured domains for app:
    =====> configtest.XXXXXXX.XXXXXXX.com
    -----> Domains found in SSL certificate:
    =====> *.XXXXXXX.com
    =====> *.XXXXXXX.com
    =====> XXXXXXX.com
    -----> Setting config vars
           DOKKU_NGINX_SSL_PORT: 443
    -----> Configuring configtest.XXXXXXX.XXXXXXX.com...(using /var/lib/dokku/plugins/available/nginx-vhosts/templates/nginx.conf.template)
    -----> Creating https nginx.conf
    -----> Running nginx-pre-reload
           Reloading nginx
    -----> Setting config vars
           DOKKU_APP_RESTORE: 1
    =====> Application deployed:
           http://configtest.XXXXXXX.XXXXXXX.com

    To dokku@XXXXXXX.XXXXXXX.com:configtest
     * [new branch]      master -> master
    ```

4. Verify the vars are set

    ```
    $ dokku config configtest
    =====> configtest config vars
    DOKKU_APP_RESTORE:    1
    DOKKU_NGINX_PORT:     80
    DOKKU_NGINX_SSL_PORT: 443
    MYBUILDPACK:          http://example.com/buildpack


    $ dokku run configtest bash
    u6394@9fcf98dce0fd:~$ cat /app/.env 
    export CURL_CONNECT_TIMEOUT='5'
    export CURL_TIMEOUT='30'
    export DOKKU_DEFAULT_CHECKS_WAIT='5'
    export DOKKU_WAIT_TO_RETIRE='5'

    export MYBUILDPACK='http://example.com/buildpack'
    ```

5. Unset the variable set using `dokku config:set` prior to the app being deployed.

    ```
    $ dokku config:unset configtest MYBUILDPACK
    -----> Unsetting MYBUILDPACK
    -----> Restarting app configtest
    -----> Releasing configtest (dokku/configtest:latest)...
    -----> Deploying configtest (dokku/configtest:latest)...
    -----> DOKKU_SCALE file found (/home/dokku/configtest/DOKKU_SCALE)
    =====> web=1
    -----> Running pre-flight checks
           For more efficient zero downtime deployments, create a file CHECKS.
           See http://dokku.viewdocs.io/dokku/checks-examples.md for examples
           CHECKS file not found in container: Running simple container check...
    -----> Waiting for 5 seconds ...
    -----> Default container check successful!
    =====> configtest container output:
            * Running on http://0.0.0.0:5000/
    =====> end configtest container output
    -----> Running post-deploy
    -----> Found previous container(s) (b19c70a55154) named configtest.web.1
    =====> renaming container (b19c70a55154) configtest.web.1 to configtest.web.1.1462987903
    =====> renaming container (31c050c92302) boring_pike to configtest.web.1
    -----> No matching configured domains for configtest found in SSL certificate. Your app will show as insecure in a browser if accessed via SSL
    -----> Please add appropriate domains via the dokku domains command
    -----> Configured domains for app:
    =====> configtest.XXXXXXX.XXXXXXX.com
    -----> Domains found in SSL certificate:
    =====> *.XXXXXXX.com
    =====> *.XXXXXXX.com
    =====> XXXXXXX.com
    -----> Configuring configtest.XXXXXXX.XXXXXXX.com...(using /var/lib/dokku/plugins/available/nginx-vhosts/templates/nginx.conf.template)
    -----> Creating https nginx.conf
    -----> Running nginx-pre-reload
           Reloading nginx
    -----> Setting config vars
           DOKKU_APP_RESTORE: 1
    -----> Shutting down old containers in 5 seconds
    =====> b19c70a551541d01d485330f0123946d3c857a2773eb6313880f1adfc8a6c40d
    =====> Application deployed:
           http://configtest.XXXXXXX.XXXXXXX.com
    ```

6. Check for vars again to see if the config var has been removed.

    ```
    $ dokku config configtest
    =====> configtest config vars
    DOKKU_APP_RESTORE:    1
    DOKKU_NGINX_PORT:     80
    DOKKU_NGINX_SSL_PORT: 443

    $ dokku run configtest bash
    u6394@8444b0a26a77:~$ cat /app/.env 
    export CURL_CONNECT_TIMEOUT='5'
    export CURL_TIMEOUT='30'
    export DOKKU_DEFAULT_CHECKS_WAIT='5'
    export DOKKU_WAIT_TO_RETIRE='5'

    export MYBUILDPACK='http://example.com/buildpack'
    ```

7. Set another var

    ```
    $ dokku config:set configtest TEST_VAR_3='TEST_VAR_3'
    -----> Setting config vars
           TEST_VAR_3: TEST_VAR_3
    -----> Restarting app configtest
    -----> Releasing configtest (dokku/configtest:latest)...
    -----> Deploying configtest (dokku/configtest:latest)...
    -----> DOKKU_SCALE file found (/home/dokku/configtest/DOKKU_SCALE)
    =====> web=1
    -----> Running pre-flight checks
           For more efficient zero downtime deployments, create a file CHECKS.
           See http://dokku.viewdocs.io/dokku/checks-examples.md for examples
           CHECKS file not found in container: Running simple container check...
    -----> Waiting for 5 seconds ...
    -----> Default container check successful!
    =====> configtest container output:
            * Running on http://0.0.0.0:5000/
    =====> end configtest container output
    -----> Running post-deploy
    -----> Found previous container(s) (31c050c92302) named configtest.web.1
    =====> renaming container (31c050c92302) configtest.web.1 to configtest.web.1.1462988177
    =====> renaming container (21e338894c2b) tiny_booth to configtest.web.1
    -----> No matching configured domains for configtest found in SSL certificate. Your app will show as insecure in a browser if accessed via SSL
    -----> Please add appropriate domains via the dokku domains command
    -----> Configured domains for app:
    =====> configtest.XXXXXXX.XXXXXXX.com
    -----> Domains found in SSL certificate:
    =====> *.XXXXXXX.com
    =====> *.XXXXXXX.com
    =====> XXXXXXX.com
    -----> Configuring configtest.XXXXXXX.XXXXXXX.com...(using /var/lib/dokku/plugins/available/nginx-vhosts/templates/nginx.conf.template)
    -----> Creating https nginx.conf
    -----> Running nginx-pre-reload
           Reloading nginx
    -----> Setting config vars
           DOKKU_APP_RESTORE: 1
    -----> Shutting down old containers in 5 seconds
    =====> 31c050c92302b35215f5d68b10470f3f74cbc8ebc808aa5e0ed841597c660c7d
    =====> Application deployed:
           http://configtest.XXXXXXX.XXXXXXX.com
    ```

8. Check if the var is set 

    ```
    $ dokku config configtest
    =====> configtest config vars
    DOKKU_APP_RESTORE:    1
    DOKKU_NGINX_PORT:     80
    DOKKU_NGINX_SSL_PORT: 443
    TEST_VAR_3:           TEST_VAR_3

    $ dokku run configtest bash
    u6394@26389f1c63db:~$ cat /app/.env 
    export CURL_CONNECT_TIMEOUT='5'
    export CURL_TIMEOUT='30'
    export DOKKU_DEFAULT_CHECKS_WAIT='5'
    export DOKKU_WAIT_TO_RETIRE='5'

    export MYBUILDPACK='http://example.com/buildpack'
    ```

9. Workaround (dokku:rebuild followed by a config:set seems to do the job)

    ```
    $ dokku ps:rebuild configtest
    -----> Cleaning up...
    -----> Building configtest from herokuish...
    -----> Injecting apt repositories and packages ...
    -----> Adding BUILD_ENV to build environment...
    -----> Python app detected
    -----> Installing dependencies with pip
    You are using pip version 7.1.2, however version 8.1.2 is available.
    You should consider upgrading via the 'pip install --upgrade pip' command.
           
    -----> Discovering process types
           Procfile declares types -> web
    -----> Releasing configtest (dokku/configtest:latest)...
    -----> Deploying configtest (dokku/configtest:latest)...
    -----> DOKKU_SCALE file found (/home/dokku/configtest/DOKKU_SCALE)
    =====> web=1
    -----> Running pre-flight checks
           For more efficient zero downtime deployments, create a file CHECKS.
           See http://dokku.viewdocs.io/dokku/checks-examples.md for examples
           CHECKS file not found in container: Running simple container check...
    -----> Waiting for 5 seconds ...
    -----> Default container check successful!
    =====> configtest container output:
            * Running on http://0.0.0.0:5000/
    =====> end configtest container output
    -----> Running post-deploy
    -----> Found previous container(s) (21e338894c2b) named configtest.web.1
    =====> renaming container (21e338894c2b) configtest.web.1 to configtest.web.1.1463021116
    =====> renaming container (3d1c0f167113) elated_heyrovsky to configtest.web.1
    -----> No matching configured domains for configtest found in SSL certificate. Your app will show as insecure in a browser if accessed via SSL
    -----> Please add appropriate domains via the dokku domains command
    -----> Configured domains for app:
    =====> configtest.XXXXXXX.XXXXXXX.com
    -----> Domains found in SSL certificate:
    =====> *.XXXXXXX.com
    =====> *.XXXXXXX.com
    =====> XXXXXXX.com
    -----> Configuring configtest.XXXXXXX.XXXXXXX.com...(using /var/lib/dokku/plugins/available/nginx-vhosts/templates/nginx.conf.template)
    -----> Creating https nginx.conf
    -----> Running nginx-pre-reload
           Reloading nginx
    -----> Setting config vars
           DOKKU_APP_RESTORE: 1
    -----> Shutting down old containers in 5 seconds
    =====> 21e338894c2b7512f1ce92f41ed0670120e81c533330483ecb0fd165badaf919
    =====> Application deployed:
           http://configtest.XXXXXXX.XXXXXXX.com

    $ dokku config configtest
    =====> configtest config vars
    DOKKU_APP_RESTORE:    1
    DOKKU_NGINX_PORT:     80
    DOKKU_NGINX_SSL_PORT: 443
    TEST_VAR_3:           TEST_VAR_3
    
    $ dokku run configtest bash
    u1511@f0fc353be6f4:~$ cat /app/.env 
    export CURL_CONNECT_TIMEOUT='5'
    export CURL_TIMEOUT='30'
    export DOKKU_DEFAULT_CHECKS_WAIT='5'
    export DOKKU_WAIT_TO_RETIRE='5'

    export DOKKU_APP_RESTORE='1'
    export DOKKU_NGINX_PORT='80'
    export DOKKU_NGINX_SSL_PORT='443'
    export TEST_VAR_3='TEST_VAR_3'
    ```

### PR DESCRIPTION
**Pull request by josegonzalez**
Closes #2177

### CODE REVIEW

**Comment by michaelshobbs** [comment]
Perhaps instead of dropping this, we make it clear that the file is used for the build time ENV only?

**Comment by josegonzalez** [comment]
Heroku doesn't support this, I don't know why we should.

**Comment by michaelshobbs** [comment]
I need to look at herokuish before we merge this. It might use `.env`. The failed test here indicates something is using it.

**Comment by michaelshobbs** [comment]
TL;DR
herokuish will need to be updated before this PR is merged.

Yup herokuish still uses this for "backwards compatibility" https://github.com/gliderlabs/herokuish/blob/v0.3.13/include/buildpack.bash#L92-L95

However, that's the only point in the build process that herokuish sources anything having to do with app environment. Thus this would break custom config of any environment variable; including `CURL_*` and more to the point of the failing test, `BUILDPACK_URL`.

**Comment by josegonzalez** [comment]
Ah interesting. So the easiest fix is a doc fix, or we could rebuild whenever the app is released?

**Comment by michaelshobbs** [comment]
docs, yes
rebuild always, no

Not everyone needs (or more importantly wants) to rebuild on restart. I personally prefer this flexibility. Again, apps should not be reading `.env` and instead use the actual `ENV` object from their language/framework. 

Third option: We could delete `.env` after the image has been built.

**Comment by josegonzalez** [comment]
What if someone writes a .env file to their repository (for whatever reason).

**Comment by michaelshobbs** [comment]
Then docs it is 😜
